x/ibc-hooks/contracts/crosschain-swaps/src/execute.rs

```diff
@@ -0,0 +1,271 @@
+use cosmwasm_std::{coins, from_binary, to_binary, wasm_execute, BankMsg, Reply, Timestamp};
+use cosmwasm_std::{Addr, Coin, DepsMut, Response, SubMsg, SubMsgResponse, SubMsgResult};
+use swaprouter::msg::{ExecuteMsg as SwapRouterExecute, Slippage, SwapResponse};
+
+use crate::checks::{validate_memo, validate_receiver};
+use crate::consts::{CALLBACK_KEY, FORWARD_REPLY_ID, PACKET_LIFETIME, SWAP_REPLY_ID};
+use crate::ibc::{MsgTransfer, MsgTransferResponse};
+use crate::msg::{CrosschainSwapResponse, Recovery};
+
+use crate::state::{
+    ForwardMsgReplyState, ForwardTo, IBCTransfer, Status, SwapMsgReplyState, CONFIG,
+    FORWARD_REPLY_STATES, INFLIGHT_PACKETS, RECOVERY_STATES, SWAP_REPLY_STATES,
+};
+use crate::ContractError;
+
+/// This is the main execute call of this contract.
+///
+/// It's objective is to trigger a swap between the supplied pairs
+///
+#[allow(clippy::too_many_arguments)]
+pub fn swap_and_forward(
+    deps: DepsMut,
+    block_time: Timestamp,
+    contract_addr: Addr,
+    input_coin: Coin,
+    output_denom: String,
+    slippage: Slippage,
+    receiver: Addr,
+    next_memo: Option<String>,
+    failed_delivery: Option<Recovery>,
+) -> Result<Response, ContractError> {
+    deps.api.debug(&format!("executing swap and forward"));
+    let config = CONFIG.load(deps.storage)?;
+
+    // Message to swap tokens in the underlying swaprouter contract
+    let swap_msg = SwapRouterExecute::Swap {
+        input_coin: input_coin.clone(),
+        output_denom,
+        slippage,
+    };
+    let msg = wasm_execute(config.swap_contract, &swap_msg, vec![input_coin])?;
+
+    // Check that the received is valid and retrieve its channel
+    let (valid_channel, valid_receiver) = validate_receiver(deps.as_ref(), receiver)?;
+    // If there is a memo, check that it is valid
+    if let Some(memo) = &next_memo {
+        validate_memo(memo)?;
+    }
```

Comment:
Perhaps good to clarify what is the needed safety check? (and maybe a one sentence for why no callback?)

```suggestion
    if let Some(memo) = &next_memo {
        let json = parse_json(memo)?;
        ensure_no_callback(json)?;
    }
```

Reply:
Cleaned this up and split into multiple single-purpose functions. Added some docs.